### PR TITLE
[WIP] Second attempt on custom environment info

### LINKED ISSUE #384: Print Vector<byte>.Count as part of machine info
Helps to know this when benchmarking SIMD operations

### PR DESCRIPTION
**Pull request by gsomix**
Closes #384 

- [x] Add `[CustomEnvironmentInfo]` attribute
- [x] Implement attribute parsing
- [x] Add invocation of custom info method to benchmark's template and code generator
- [ ] Summary
- [ ] Validator, tests, docs

### CODE REVIEW

File: src/BenchmarkDotNet/Code/CodeGenerator.cs [hunk 1]
```diff
@@ -95,6 +97,24 @@ internal static string Generate(BuildPartition buildPartition)
             return benchmarkProgramContent;
         }
 
+        private static string GetCustomEnvInfo(DeclarationsProvider provider)
+        {
+            if (provider.CustomEnvInfoMethodName is null)
+            {
+                return "Array.Empty<string>()";
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Code/CodeGenerator.cs]: Could you please use the full name? 

```suggestion
                return "System.Array.Empty<string>()";
```

File: src/BenchmarkDotNet/Code/CodeGenerator.cs [hunk 2]
```diff
@@ -95,6 +97,24 @@ internal static string Generate(BuildPartition buildPartition)
...
+            }
+
+            switch (provider.CustomEnvInfoMethodReturnType)
+            {
+                case Type t when t == typeof(string):
+                    return $"new string[] {{ {provider.CustomEnvInfoMethodName}() }}";
+                case Type t when typeof(IEnumerable<string>).IsAssignableFrom(t):
+                    return $"{provider.CustomEnvInfoMethodName}()";
+                default:
+                    return "Array.Empty<string>()";
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Code/CodeGenerator.cs]: Could you please use the full name?

```suggestion
                    return "System.Array.Empty<string>()";
```

File: src/BenchmarkDotNet/Code/CodeGenerator.cs (hunk 1 above)
Comment (gsomix) [reply src/BenchmarkDotNet/Code/CodeGenerator.cs]: OK, let's try this suggestons feature. :)

**Comment by gsomix** [comment]
@adamsitnik  I'm trying to figure out where to put environment info in the report. There are 3 possible solutions:

- add info to summary. I don't like this solution because summary should be clear and concise.
- print info in detailed reports for every job. It's better, but harder to discover.
- introduce new conclusion for environment info

What do you think? Thanks!

**Comment by adamsitnik** [comment]
@gsomix maybe a new column? @AndreyAkinshin what do you think?

**Comment by gsomix** [comment]
@adamsitnik New column makes sense only for short info, e.g. `Vector<byte>.Count`. I think we can do that and just drop support of `IEnumerable<string>` version at all.

**Comment by adamsitnik** [comment]
@gsomix good idea! if users want more info, they can always concatenate the strings

**Comment by gsomix** [comment]
Guys, sorry for the delya. I'm going back to work.

**Comment by roji** [comment]
@gsomix any plans to continue work on this?

**Comment by gsomix** [comment]
@roji Yeah, sorry. I'm going to take a fresh look at PR this weekend.

**Comment by gsomix** [comment]
@roji Sorry, I cannot find a time for it, I'm really busy at work last days. I'm going to close it, but feel free rebase it and push to merge.

**Comment by roji** [comment]
Thanks for your message @gsomix, I'll try to pick this up in the coming days.

**Comment by MarekM25** [comment]
@adamsitnik, @gsomix 
I would like to verify my understanding of this issue. Please, correct me if I am wrong. So, the expected summary is:
``` ini

BenchmarkDotNet=v0.11.5.20191006-develop, OS=Windows 10.0.17763.775 (1809/October2018Update/Redstone5)
Intel Core i5-8250U CPU 1.60GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
  [Host] : .NET Framework 4.7.2 (4.7.3460.0), X64 LegacyJIT DEBUG  [AttachedDebugger]
  Dry    : .NET Framework 4.7.2 (4.7.3460.0), X64 RyuJIT

Job=Dry  IterationCount=1  LaunchCount=1  
RunStrategy=ColdStart  UnrollFactor=1  WarmupCount=1  

```
|           Method |     Mean | Error | Custom environment info |
|----------------- |---------:|------:|------------------------:|
|            Sleep | 24.51 ms |    NA | Vector<byte>.Count = 32 |
| Thread.Sleep(10) | 16.62 ms |    NA | Vector<byte>.Count = 32 |

I plan to add it to a summary by modifying _MethodsToBenchmarksWithFullConfig_ in _BenchmarkConverter_ class. There I will add a new column to config by executing customEnvInfoMethods added by @gsomix .

**Comment by gsomix** [comment]
@MarekM25 @roji Sorry for the late answer. It's a difficult question. Indeed the environment variables cannot vary according to benchmark methods, but consider benchmarking against different Jobs. Personally I'd prefer to place that information into dedicated column, because along with grouping by methods it's more descriptive way than putting additional lines to summary header. But I'm not sure.

@adamsitnik @AndreyAkinshin What do you think guys?

**Comment by MarekM25** [comment]
@roji @gsomix 
I also prefer the first idea. For me, it is more natural and elegant. For example:
``` ini

BenchmarkDotNet=v0.11.5.20191006-develop, OS=Windows 10.0.17763.775 (1809/October2018Update/Redstone5)
Intel Core i5-8250U CPU 1.60GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
  [Host] : .NET Framework 4.7.2 (4.7.3460.0), X64 LegacyJIT DEBUG  [AttachedDebugger]
  Dry    : .NET Framework 4.7.2 (4.7.3460.0), X64 RyuJIT

Job=Dry  IterationCount=1  LaunchCount=1  
RunStrategy=ColdStart  UnrollFactor=1  WarmupCount=1  
Custom environment info: Vector<byte>.Count = 32
```
If it is about implementation, I propose to use static HostEnvironmentInfo class and there add a string property for CustomEnvInfo. We will have one custom info for all executed benchmark classes.

**Comment by gsomix** [comment]
@MarekM25 I'm not sure that's good idea, because we might want to actually invoke methods to retrieve information about environment from generated code. So it may be different from host environment variables and also may vary according to different Runtimes/Jobs.

Please read this comment from Adam in first PR for this issue.
https://github.com/dotnet/BenchmarkDotNet/pull/893#pullrequestreview-164559842

Thanks for working on it!

**Comment by MarekM25** [comment]
@gsomix 
Do you have some implementation tips for it? I see that you did it in generated code, but how to do that in summary? I mean not by adding new column but below environment info with vary for different Runtimes/Jobs.

**Pull request closed without merging**